src/shared_memory_manager.h

```diff
@@ -130,31 +106,71 @@ class SharedMemoryManager {
   TRITONSERVER_Error* UnregisterAll(TRITONSERVER_MemoryType memory_type);
 
+#ifdef TRITON_ENABLE_GPU
+  /// Add a shared memory block representing shared memory in CUDA
+  /// (GPU) memory to the manager. Return TRITONSERVER_ERROR_ALREADY_EXISTS
+  /// if a shared memory block of the same name already exists in the manager.
+  /// \param name The name of the memory block.
+  /// \param cuda_shm_handle The unique memory handle to the cuda shared
+  /// memory block.
+  /// \param byte_size The size, in bytes of the block.
+  /// \param device id The GPU number the shared memory region is in.
+  /// \return a TRITONSERVER_Error indicating success or failure.
+  TRITONSERVER_Error* RegisterCUDASharedMemory(
+      const std::string& name, const cudaIpcMemHandle_t* cuda_shm_handle,
+      const size_t byte_size, const int device_id);
+
+  /// Get the CUDA memory handle associated with the block name.
+  /// Return TRITONSERVER_ERROR_NOT_FOUND if named block doesn't exist.
+  /// \param name The name of the shared memory block to get.
+  /// \param cuda_mem_handle Returns the cuda memory handle with the memory
+  /// block.
+  /// \return a TRITONSERVER_Error indicating success or failure.
+  TRITONSERVER_Error* GetCUDAHandle(
+      const std::string& name, cudaIpcMemHandle_t** cuda_mem_handle);
+#endif  // TRITON_ENABLE_GPU
+
  private:
   /// A helper function to remove the named shared memory blocks of
   /// specified type
   TRITONSERVER_Error* UnregisterHelper(
       const std::string& name, TRITONSERVER_MemoryType memory_type);
 
+  struct ShmFile {
+#ifdef _WIN32
+    HANDLE shm_file_;
+    ShmFile(void* shm_file) { shm_file_ = static_cast<HANDLE>(shm_file); };
+    HANDLE GetShmFile() { return shm_file_; };
+#else
+    std::unique_ptr<int> shm_file_;
+    ShmFile(void* shm_file)
+    {
+      shm_file_ = std::make_unique<int>(*static_cast<int*>(shm_file));
+    };
+    int* GetShmFile() { return shm_file_.get(); };
+#endif  // _WIN32
+  };
+
   /// A struct that records the shared memory regions registered by the shared
   /// memory manager.
   struct SharedMemoryInfo {
     SharedMemoryInfo(
         const std::string& name, const std::string& shm_key,
-        const size_t offset, const size_t byte_size, int shm_fd,
+        const size_t offset, const size_t byte_size, void* shm_file,
         void* mapped_addr, const TRITONSERVER_MemoryType kind,
         const int64_t device_id)
         : name_(name), shm_key_(shm_key), offset_(offset),
-          byte_size_(byte_size), shm_fd_(shm_fd), mapped_addr_(mapped_addr),
-          kind_(kind), device_id_(device_id)
+          byte_size_(byte_size),
+          platform_handle_(std::make_unique<ShmFile>(shm_file)),
+          mapped_addr_(mapped_addr), kind_(kind), device_id_(device_id)
     {
     }
 
     std::string name_;
     std::string shm_key_;
     size_t offset_;
     size_t byte_size_;
-    int shm_fd_;
+    std::unique_ptr<ShmFile> platform_handle_;
```

Comment:
I didn't get this far down when I wrote the above comment. We are wrapping this struct in a `unique_ptr` anyway so why does the underlying data structure need to be a smart pointer? I am making a lot of rhetorical questions, but it's always asking what value are we getting out of using this more complicated thing?

Again, I would say this could be converted into a raw pointer with no impact as long as we free it in the destructor or whenever it needs to be freed. After looking at the code, `SharedMemoryInfo` isn't the one owning the pointer, the `shared_memory_map_` really is since this lives as long as the entry in the map does. Directly below, `mapped_addr_` would cause a memory leak if we didn't clean up the shared memory regions properly, but AFAIK it doesn't so we should be fine.

Reply:
I think this is mostly to do with the safest possible design. The current main branch code in client uses raw pointers and leaks memory because it's easy to forget to release the memory. Using the unique_ptrs, what we lose in additional verbosity, we gain in not needing to update a destructor and being confident memory will be safely freed.

I think this will be particularly helpful when/if ShmFile becomes a more complex data structure.

Reply:
Discussed offline. Decision was to change the structure of ShmFile, but to continue using smart pointers.